### DISCUSSION TITLE
One Week project ideas

### DISCUSSION THREAD

**Discussion by AAkindele** [retaildevcrews/ngsa#724] [General]
This is a list of ideas for One Week projects. Feel free to add and update.

**Comment by AAkindele** [comment]
Title: Auto-generate GitOps template for an application.

Description:

On the initial GitOps setup of an application, an application developer needs to create the config files on their own. Create a tool that allows a developer to generate GitOps config files for their application.

Impact:

- Improve the initial developer experience of setting up GitOps for their application.
- Reduce the time it takes an application developer to deploy their app.

Details:

- This can be a script or a CLI tool.
- The default values allows a developer to check in the files as is and deploy a sample app to a target cluster

**Comment by AAkindele** [comment]
Title: Static analysis of Kubernetes config in CI/CD pipeline.

Description:

When running AKS in production, there are situations where a valid kubernetes yaml definition cannot, or should not, be applied to a cluster. Reasons range from Azure policies, RBAC configs, specialized hardware requirements. Detect these issues early in the CI/CD pipeline before they fail in the cluster.

Impact:

- Improve feedback loop for app developer when debugging deployments.
- Platform team can rollout early detection to all app teams easily from centralized location.

Details:

- Checkov: <https://github.com/bridgecrewio/checkov>
  - written in python
  - comes with some built-in validators
  - can write custom validators in python
- Gatekeeper: <https://github.com/open-policy-agent/gatekeeper>
  - powers Azure policies on AKS
  - check yaml files against target cluster policies before they reach the cluster
  - GCP example with cloud build and Anthos: <https://cloud.google.com/anthos-config-management/docs/tutorials/app-policy-validation-ci-pipeline>
    - Main tools for validation: K8s constraints, K8s constrainttemplates, gcr.io/kpt-dev/kpt, gcr.io/kpt-functions/gatekeeper-validate
    - <https://googlecontainertools.github.io/kpt/reference/fn/source/>
    - <https://googlecontainertools.github.io/kpt/guides/consumer/function/catalog/validators/>

**Comment by AAkindele** [comment]
Title: Multi-cluster deployment consistency.

Description:

In a multi-cluster environment, application deployments can fail in a subset of the target clusters. This leads to a situation where different versions of the application are running in the environment. Detect deployment failures.

Impact:

- Improve

Details:

- Detect deployment failures in a multi-cluster GitOps environment.
- Rollback all clusters to previous version on partial deployment failures.
- The partial deployment failures and roll back are observable are the application developer.
- Bonus? Configurable if platform should rollback or only raise a warning?

**Comment by AAkindele** [comment]
Title: Expand on code server and debugging tools for development cluster.

Description:

- TODO

Impact:

- TODO

Details:

- TODO

**Comment by jofultz** [comment]
Title: Create a VS Code Extension that will provide minimal validation of kitt or gitops yaml files.

**Comment by atxryan** [comment]
Title: Monitoring: deploying CNCF *on* the cluster

Description:

- TODO

Impact:

- TODO

Details:

- TODO

**Comment by atxryan** [comment]
Title: Isolated traffic control

Description:

- The default configuration of Traefik uses a shared traffic model where all application traffic is routed to a node then to a pod.  The issue is the application pod may not be running on that node and may need to take an extra hop to a node where the pod is running.  The goal is to configure the upstream load balance with only nodes that are running the application pods so the traffic is guaranteed to be routed to an appropriate node.

Impact:

- In high performance applications, this extra hop could negate all the performance tuning work done on the application in the overall performance equation.

Details:

- TODO

**Comment by atxryan** [comment]
Title: Arc 

Description:

- TODO

Impact:

- TODO

Details:

- TODO
